File: shell/internal/adapters/testsupport/sqlite.go
package testsupport

import (
	"database/sql"
	"os"
	"path/filepath"
	"testing"
	"time"

	_ "modernc.org/sqlite"

	"github.com/kirathecat/kira-studio/shell/internal/storage/model"
)

const bigRowsSqlite = 1_000_000

// SqliteFixture is sqlite.ts's SqliteFixture — D32: a temp-file fixture, not a container. No image
// pin, no healthcheck, no wait strategy, no root-versus-app-user split, no resolveDockerHost() — a
// file-based engine makes this the simplest fixture of the four.
type SqliteFixture struct {
	Path   string
	Dir    string
	Config model.ResolvedConnectionConfig
}

var sqliteMemo fixture[SqliteFixture]

const sqliteDatabaseFile = "kira_test.sqlite"

// StartSqlite is sqlite.ts's startSqlite. No Docker gate — nothing it needs depends on the daemon.
// Memoized the same way every other fixture in this package is (B15): termination happens once,
// from the package's own TestMain, never from an individual test's t.Cleanup.
func StartSqlite(t *testing.T) *SqliteFixture {
	t.Helper()
	f, err := sqliteMemo.get(startSqlite)
	if err != nil {
		t.Fatalf("sqlite fixture: %v", err)
	}
	return f
}

// StopSqlite removes the memoized temp directory, if one was ever created. Call once, from the
// test binary's own TestMain, after m.Run() returns.
func StopSqlite() {
	sqliteMemo.stop(func(f *SqliteFixture) { _ = os.RemoveAll(f.Dir) })
}

func startSqlite() (*SqliteFixture, error) {
	dir, err := os.MkdirTemp("", "kira-sqlite-")
	if err != nil {
		return nil, err
	}
	path := filepath.Join(dir, sqliteDatabaseFile)

	seedPath := filepath.Join(repoRoot(), "tests", "db", "fixtures", "0009_sqlite_seed.sql")
	seedSQL, err := os.ReadFile(seedPath)
	if err != nil {
		return nil, err
	}

	// Plain create-if-missing DSN, deliberately not the adapter's own mode=ro/mode=rw — this is the
	// fixture creating the file for the first time, the one case in the whole codebase where SQLite
	// creating a database on open is exactly what is wanted (D8's "Kira never creates a database"
	// rule is about the *adapter*, not this harness).
	db, err := sql.Open("sqlite", "file:"+path)
	if err != nil {
		return nil, err
	}
	defer db.Close()

	if _, err := db.Exec(string(seedSQL)); err != nil {
		return nil, err
	}
	// One plain WITH RECURSIVE CTE inserts all 1,000,000 rows directly — no chunking, no PRAGMA
	// adjustment needed, unlike MariaDB's SEQUENCE-engine seq_1_to_N or MySQL's chunked digits cross
	// join (sqlite.ts's own comment, ported verbatim).
	bigRowsSQL := `WITH RECURSIVE seq(n) AS (
		SELECT 1
		UNION ALL
		SELECT n + 1 FROM seq WHERE n < ` + itoaPositive(bigRowsSqlite) + `
	)
	INSERT INTO big_rows (id, payload)
	SELECT n, hex(randomblob(16)) FROM seq`
	if _, err := db.Exec(bigRowsSQL); err != nil {
		return nil, err
	}
	// ANALYZE on big_rows only, mirroring 0002_mariadb_seed.sql's own note — every other table is
	// left with no sqlite_stat1 row, which is what scenario 6 needs to assert a null estimate.
	if _, err := db.Exec("ANALYZE big_rows"); err != nil {
		return nil, err
	}

	now := time.Now().UTC().Format(time.RFC3339Nano)
	cfg := model.ResolvedConnectionConfig{
		ID: "test-sqlite", SortOrder: 0, CreatedAt: now, UpdatedAt: now,
		Name: "Test SQLite", Kind: "sqlite", Color: "violet", Mode: "fields", ReadOnly: false,
		Database: Strp(path), Options: map[string]any{},
	}
	return &SqliteFixture{Path: path, Dir: dir, Config: cfg}, nil
}

func itoaPositive(n int) string {
	if n == 0 {
		return "0"
	}
	var buf [20]byte
	i := len(buf)
	for n > 0 {
		i--
		buf[i] = byte('0' + n%10)
		n /= 10
	}
	return string(buf[i:])
}

File: packages/db-fixtures/support/sqlite.ts
import { mkdtempSync, readFileSync, rmSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join, resolve } from 'node:path';
import type { ResolvedConnectionConfig } from './connectionConfig';

const BIG_ROWS = 1_000_000;
const SEED_SQL_PATH = resolve(__dirname, '../fixtures/0009_sqlite_seed.sql');
const DATABASE_FILE = 'kira_test.sqlite';

export interface SqliteFixture {
  path: string;
  /** The temp directory `path` lives in — sibling paths for a missing-file/directory/roundtrip
   *  connect-failure test live here too, rather than each test managing its own mkdtemp. */
  dir: string;
  config: ResolvedConnectionConfig; // ready to hand to the adapter
  stop(): Promise<void>;
}

// D33: this suite's one environment dependency — everything else it needs (a temp directory, a
// file) runs anywhere. Mirrors support/docker.ts's isDockerAvailable()/DOCKER_UNAVAILABLE_MESSAGE
// discipline: a legible skip naming the runtime requirement, not a bare module-resolution error.
export async function sqliteAvailable(): Promise<boolean> {
  try {
    await import('node:sqlite');
    return true;
  } catch {
    return false;
  }
}

export const SQLITE_UNAVAILABLE_MESSAGE =
  'node:sqlite is unavailable in this runtime — Bun 1.4+ (or Electron/Node 22.5+) is required ' +
  'to run packages/db-fixtures/sqlite.spec.ts.';

// One temp database per test process, same discipline as support/mariadb.ts (§11b) — though here
// it buys nothing but the same memoization shape, since there is no container to reuse.
let memoized: Promise<SqliteFixture> | null = null;

export function startSqlite(opts?: { seedBigTable?: boolean }): Promise<SqliteFixture> {
  if (!memoized) memoized = start(opts);
  return memoized;
}

async function start(opts?: { seedBigTable?: boolean }): Promise<SqliteFixture> {
  // D32: a temp-file fixture, not a container — no image pin, no healthcheck, no wait strategy,
  // no root-versus-app-user split, no resolveDockerHost(). A file-based engine makes this the
  // simplest fixture of the eight.
  const { DatabaseSync } = await import('node:sqlite');
  const dir = mkdtempSync(join(tmpdir(), 'kira-sqlite-'));
  const path = join(dir, DATABASE_FILE);

  const db = new DatabaseSync(path);
  try {
    db.exec(readFileSync(SEED_SQL_PATH, 'utf8'));

    // Gated: the harness only needs to prove it can build the table; P2's paging is what
    // actually consumes 1M rows. Default true, unlike P1, because P2 genuinely consumes them —
    // same rule support/mariadb.ts's own gate follows.
    //
    // Unlike MariaDB's SEQUENCE-engine seq_1_to_N or MySQL's chunked digits cross join, one plain
    // WITH RECURSIVE CTE inserts all 1,000,000 rows directly (~1s, F11) — no chunking, no PRAGMA
    // adjustment needed.
    if (opts?.seedBigTable ?? true) {
      db.exec(
        `WITH RECURSIVE seq(n) AS (
           SELECT 1
           UNION ALL
           SELECT n + 1 FROM seq WHERE n < ${BIG_ROWS}
         )
         INSERT INTO big_rows (id, payload)
         SELECT n, hex(randomblob(16)) FROM seq`,
      );
      // ANALYZE on big_rows only, mirroring 0002_mariadb_seed.sql's own note — every other table
      // is left with no sqlite_stat1 row, which is what scenario 6 needs to assert a null estimate.
      db.exec('ANALYZE big_rows');
    }
  } finally {
    db.close();
  }

  const now = new Date().toISOString();
  const config: ResolvedConnectionConfig = {
    id: 'test-sqlite',
    sortOrder: 0,
    createdAt: now,
    updatedAt: now,
    name: 'Test SQLite',
    kind: 'sqlite',
    color: 'violet',
    mode: 'fields',
    readOnly: false,
    host: null,
    port: null,
    database: path,
    username: null,
    uri: null,
    options: {},
    password: null,
  };

  return {
    path,
    dir,
    config,
    async stop() {
      // Playwright's workers:1 config runs every UI spec file sequentially in the same worker
      // process, sharing this module's state — support/mariadb.ts's own reason for resetting the
      // memo before the next spec file's startSqlite() call.
      memoized = null;
      rmSync(dir, { recursive: true, force: true });
    },
  };
}
